core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1119,6 +1119,15 @@ class ReplicaManager(val config: KafkaConfig,
         else
           Set.empty[Partition]
 
+        /*
+         * KAFKA-8392
+         * For topic partitions of which the broker is no longer a leader, delete metrics (bytesIn, bytesOut, messagesIn)
+         * related to those topics. Note that this means the broker stops being either a replica or a leader of
+         * partitions of said topics
+         */
+        val leaderTopicSet = leaderPartitionsIterator.map(_.topic).toSet
+        partitionsBecomeFollower.map(_.topic).diff(leaderTopicSet).foreach(brokerTopicStats.removeOldLeaderMetrics)
```

Comment:
(1) If a broker is only the follower for a topic, we will hit this code before the corresponding topic stat is created (this is lazily created when records are written and read). So, in that case, the topic stat is not present at this present. Later on, the topic stat will be added and there is no logic to remove the unneeded metrics. 

(2) We probably want to convert partitionsBecomeFollower.map(_.topic) to a set before calling diff.

Reply:
(1) If a broker is only the follower for a topic, then `bytesOutRate` and `replicationBytesOutRate` will not be created until it becomes a leader for some topics. If my assumption is right, don't we only need to remove the unneeded metrics when it stops becoming the leader and the logic for that will be executed when there is a `leaderAndIsrRequest`?

(2) Added the fix. Thank you!

Reply:
We've changed `messagesInRate` / `bytesInRate` inside the topic stat to be also lazily created: i.e. even if the topic stat object is created, these three metrics would not be created until they are being recorded. So if a broker is only follower of a topic, it would not create the above three metrics.